### DISCUSSION TITLE
How does charge threshold work?

### DISCUSSION THREAD

**Discussion by heikothole** [flobz/psa_car_controller#317] [Q&A]
Hi,
I have ordered a Corsa-e but I'm still waiting for it.
I want to limit the charge threshold to 80% while charging at home.

How does this python script work? Is there an API from PSA which supports this feature or polls the script the actual SOC from the car at stops charging if the threshold is reached?
Is the set threshold saved and will be used every time I connect the car to the wallbox (or any other charger) or will the set threshold only be used for the actual charging process?
I'm worried that I might want to charge to 100% when I'm on the road, and then the script is not accessible.

Thanks and best regards,
Heiko

**Comment by marcodallago** [comment]
No worries, I have the same car. You can set up a script that when the sensor of the car charging states "connected" you enable the switch that limits automatically to 80%. I've done that using Homekit at the end, which for me is more convenient. But you can do it in home assistant

**Comment by UeliF** [comment]
Hi, 
if I set the threshold to 80 %, it does work for the first time. However, if I charge again (~1week later) the car charged up to 100 %. The charge_control still reports 80% though. Not clear for me why it doesnt stick to 80 % when charging. 
As a workaround I switch threshold to 100% and then back to 80 % some short time before charging starts.

**Comment by pdhoogh** [comment]
By putting the car in delayed charge mode once the charge threshold is reached.

The PSA car controller uses the same interface to Stellantis servers as your smartphone app does. So its control options are somewhat limited. It asks for the car's status through the Stellantis servers over the cellular network. It parses the response it gets and a piece of that response is the traction battery level. Then the code in the psa controller app compares it to the locally stored battery level limit. You already know how the limit is set. When it reaches the limit, it does the only thing it has at its disposal to stop the charge: put the car in delayed charging mode. _But that means that once the delayed charging time is reached, the car will resume charging to 100%_. 

The psa car controller puts the car in delayed charging mode using the cellular network. So one would think that even if you are at a fast DC charger, the car should stop at the set level. I have not seen that yet but I must admit I rarely use the fast chargers, so I cannot confirm or deny that assumption.

One thing is sure: when you are away from home, chances are you will not be able to change the charge limit since that happens over your home network and it may not be accessible from outside your home. But that is not a problem. If the car was set to delayed charge by the psa car controller, simply pull the charge plug out and put it back in, the car will resume charging beyond the set limit. I tried that and I can confirm it works.

The method used by the psa car controller does not always work. I have seen more than once Stellantis servers are not cooperating and then nothing happens. No data comes in and no commands go out. I've seen that more than once.

There are advantages and disadvantages to the method used here, but in all it is a good stop gap for the sorely missing charge limit in the car's own system. I would hope at some point Stellantis will add that to the system, it is in their own interest since they have an 8 year battery warranty and the battery may be more reusable as a "powerwall" after having been removed from the car.

**Reply by willenic** [reply]
Hmmm, I set my charge threshold to 72% last night, the car was on defferred charge at 23:30 when it started charging. By 05:00 this morning it was at 100%. It feels like the threshold and communication is a bit flaky.

**Reply by pdhoogh** [reply]
Yep. It is getting worse. With mine it is a surprise if it _does_ work.

**Reply by UeliF** [reply]
For me it looks like the charge level never gets updated from the Stellantis servers anymore during charging. With that, there is no chance to stop charging once a specific level is reached.

The charge level only gets updated when 100 is reached, or probably when charging has stopped.

**Reply by willenic** [reply]
Not a criticism of this app, just unbelievable the functionality isn't part of an electric vehicle. Just swapped my leaf for a peugeot e208. On the leaf you could set 2 charging windows from the car. On the peugeot you can set a time to start charging but none to stop. Despite what the FAQs state. I'd laugh it wasn't costing me money.

Incidentally has anyone tried the start/stop charge command? Might try that and pair up with a target rate sensor on my tariff...

**Reply by pdhoogh** [reply]
I use an EVSE that has a timer. You can also limit the energy (on the EVSE) in one charge, it stops after e.g. 20 kWh.

I believe there is experience that the 12 Volt battery gets knackered if the car needs to communicate all the time. That may be why Stellantis is limiting comms.

I fully agree the in-car SoC limit is a key thing that is missing on the e-208.

**Reply by willenic** [reply]
Yeah was hoping to avoid shelling out another £500 with a bit of automation, not quite given up yet. Will play with the charge toggle and see what happens, otherwise I'll have to get an Ohme HomePro or something.

**Reply by marcodallago** [reply]
Tado made a new app called Smart Charging. It connects for free to psa servers /wall box app and it manages charging % limit

**Reply by pdhoogh** [reply]
It is the Stellantis servers that are failing. So Tado won't work any better than my OpenEVSE paired to the Flobz PsA car controller. Have a look at Mr. EVs trip to Italy with the Peugeot e-208: Tronity also fails... it is all using the Stellantis server data and it just isn't reliable, getting worse actually.

**Reply by UeliF** [reply]
It looks like the SoC is again being provided arount every 15 mins even during charging, this was not the case for a while and the charge threshold did not work therefore (as SoC was provided only when 100% was reached). That's nice.

**Reply by pauborrasblanco** [reply]
PSA Servers have been a mess for the last few weeks. This app used to work perfectly fine until the servers stopped working properly. The PSA official apps are now useless due to the lack of comms with the servers, and so does this project unfortunately. No choice but keep waiting...

**Comment by vineethktpla** [comment]
I agree that psa servers won't update battery level so often even during charging. A workaround would me to create an automation in the HA to stop the charge of the battery level is above 80. Charging won't stop exactly at 80 now but when psa server update the battery level it might be 85. But still the automation would stop the charge and it won't top up till 100

**Reply by 5kat3R** [reply]
I have created a template to reduce the limit by 2%, unless I have selected 100%.
```
- sensor:
  - name: modified_charge_limit
    state: >
      {%set percentage = states("input_number.charge_level") | int %}
      {% if percentage < 100 %}
        {{ percentage - 2 | float }}
      {% else %}
        {{ percentage | float }}
      {% endif%} 
```

**Comment by RTTTC** [comment]
Or simply use evcc for charging, it manages it all.
https://evcc.io/en/
